File: assistant_crm/report/omnichannel_interaction_analysis/omnichannel_interaction_analysis.js
/**
 * Omnichannel Interaction Analysis Frontend
 */

frappe.query_reports["Omnichannel Interaction Analysis"] = {
    default_print_format: "Assistant CRM - Omnichannel Interaction",
    filters: [
        {
            fieldname: "period_type",
            label: __("Period Type"),
            fieldtype: "Select",
            options: assistant_crm.report_utils.period_type_options,
            default: "Weekly",
            reqd: 1,
            on_change: function () {
                assistant_crm.report_utils.handle_period_change();
            }
        },
        {
            fieldname: "date_from",
            label: __("From Date"),
            fieldtype: "Date",
            default: frappe.datetime.add_days(frappe.datetime.get_today(), -6),
            reqd: 1
        },
        {
            fieldname: "date_to",
            label: __("To Date"),
            fieldtype: "Date",
            default: frappe.datetime.get_today(),
            reqd: 1
        }
    ],

    formatter: function (value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);

        if (column.fieldname === "channel" && data) {
            let channel = (data.channel || "").toLowerCase();
            let icon = "fa-comment";
            if (channel === "whatsapp") icon = "fa-whatsapp";
            else if (channel === "facebook") icon = "fa-facebook";
            else if (channel === "telegram") icon = "fa-paper-plane";
            else if (channel === "email") icon = "fa-envelope";

            value = `<i class="fa ${icon} mr-2"></i> ${value}`;
        }

        return value;
    }
};

File: assistant_crm/report/omnichannel_interaction_analysis/omnichannel_interaction_analysis.py
"""
Omnichannel Interaction Analysis - Script Report

Tracks conversations across all integrated channels:
- WhatsApp, Facebook, Telegram, Email, Web
- Response times and resolution status
- Channel-wise volume breakdown
"""

import json
from datetime import date
from typing import Any, Dict, List, Optional, Tuple

import frappe
from frappe.utils import getdate, flt
from assistant_crm.report.report_utils import get_period_dates

def execute(filters: Optional[Dict[str, Any]] = None) -> Tuple:
    filters = frappe._dict(filters or {})
    get_period_dates(filters)

    columns = get_columns()
    data, metrics = get_data(filters)
    chart = get_chart_data(metrics)
    report_summary = get_report_summary(metrics)

    return columns, data, None, chart, report_summary, False

def get_columns() -> List[Dict[str, Any]]:
    return [
        {"fieldname": "channel", "label": "Channel", "fieldtype": "Data", "width": 120},
        {"fieldname": "interaction_id", "label": "Interaction ID", "fieldtype": "Link", "options": "Unified Inbox Conversation", "width": 160},
        {"fieldname": "response_time", "label": "Resp. Time (Min)", "fieldtype": "Int", "width": 120},
        {"fieldname": "agent", "label": "Agent Assigned", "fieldtype": "Link", "options": "User", "width": 150},
        {"fieldname": "status", "label": "Resolution Status", "fieldtype": "Select", "width": 140},
        {"fieldname": "sla_status", "label": "SLA Status", "fieldtype": "Data", "width": 120},
    ]

def get_data(filters: frappe._dict) -> Tuple[List[Dict[str, Any]], Dict[str, Any]]:
    df = getdate(filters.date_from)
    dt = getdate(filters.date_to)

    # Fetch Unified Inbox Conversations
    conversations = frappe.get_all(
        "Unified Inbox Conversation",
        filters={"creation": [">=", df], "creation": ["<=", dt]},
        fields=["name", "platform", "agent", "status", "creation", "modified"]
    )

    data = []
    channel_counts = {}
    metrics = {"total": 0, "resolved": 0, "avg_resp": 15.0} # Mocked avg for now

    for c in conversations:
        channel = c.platform or "Web"
        channel_counts[channel] = channel_counts.get(channel, 0) + 1
        metrics["total"] += 1
        
        if c.status == "Resolved":
            metrics["resolved"] += 1

        data.append({
            "channel": channel,
            "interaction_id": c.name,
            "response_time": 5, # Demo value
            "agent": c.agent or "Unassigned",
            "status": c.status or "Open",
            "sla_status": "Met" if c.status == "Resolved" else "Pending"
        })

    metrics["channel_counts"] = channel_counts
    return data, metrics

def get_chart_data(metrics: Dict[str, Any]) -> Dict[str, Any]:
    labels = list(metrics["channel_counts"].keys())
    values = list(metrics["channel_counts"].values())
    
    return {
        "data": {
            "labels": labels,
            "datasets": [{
                "name": "Interactions",
                "values": values
            }]
        },
        "type": "pie",
        "colors": ["#1abc9c", "#3498db", "#9b59b6", "#f1c40f", "#e67e22"]
    }

def get_report_summary(metrics: Dict[str, Any]) -> List[Dict[str, Any]]:
    res_rate = round(metrics["resolved"] / metrics["total"] * 100, 1) if metrics["total"] else 0
    return [
        {"value": metrics["total"], "label": "Total Interactions", "indicator": "blue"},
        {"value": f"{res_rate}%", "label": "Resolution Rate", "indicator": "green"},
        {"value": f"{metrics['avg_resp']}m", "label": "Avg Response Time", "indicator": "yellow"}
    ]
